app/controllers/roles_controller.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class RolesController < ApplicationController
+  before_action :set_user
+  before_action :set_item, only: [:create]
+
+  def create
+    if @item && @user.has_role?(params[:role], @item)
+      redirect_back(fallback_location: root_path, flash: { alert: "User: #{@user.uid} is already assigned as #{params[:role]}" })
+    elsif @item
+      # each user only gets one role per item, remove all others first.
+      # This removes the role for all users who have that role, not only @user
+      # @user.roles.where(resource: @item).destroy_all
+
+      # Current work around, needs refactor
+      params[:role] == "viewer" ? @user.remove_role(:editor, @item) : @user.remove_role(:viewer, @item)
+      @user.add_role(params[:role], @item)
+      redirect_back(fallback_location: root_path, notice: show_notice(@user, params[:role]))
+    else
+      @user.add_role(params[:role])
+      redirect_back(fallback_location: root_path, notice: show_notice(@user, params[:role]))
+    end
+  end
+
+  # We tried using the standard destroy action but again, passing in the role id destroys the role for all users who are sharing that role, not just the user whose role we want to destroy
```

Comment:
The Rails standard destroy action of destroying a Role is the same problem we are seeing above. Users who are editors all share the same editor role. So destroying that role destroys the role for all users.